File: src/tools/system/CreateFolderTool.ts
import { PermissionLevel } from "@/types/permissions";
import type { DeviceTool } from "@/types/tools";
import { validateDeviceRelativePath } from "./devicePathValidation";

/**
 * Creates a new folder on the target device, restricted to the Agent's own
 * FileAccessPolicy allowlist (Desktop, Documents, Downloads, the Jarvis
 * folder). SAFE_ACTION: reversible (delete the folder), refuses to
 * overwrite anything that already exists.
 */
export const createFolderTool: DeviceTool = {
  id: "CREATE_FOLDER",
  name: "create_folder",
  description: "Creates a new folder on the target device, inside Desktop, Documents, Downloads, or the Jarvis folder.",
  inputSchema: {
    type: "object",
    properties: {
      path: { type: "string", description: "Path of the folder to create, e.g. 'Desktop/New Folder'." },
      deviceId: {
        type: "string",
        description: "Device to create the folder on. Defaults to the primary device if omitted.",
      },
    },
    required: ["path"],
  },
  requiredPermission: PermissionLevel.SAFE_ACTION,
  target: "device",
  validateInput(input) {
    return validateDeviceRelativePath(typeof input.path === "string" ? input.path : "");
  },
};

File: agents/imac/JarvisAgent/Sources/JarvisAgent/Tools/CreateFolderTool.swift
// REQUIRES REAL macOS VALIDATION — file system access cannot be
// exercised outside a real macOS runtime. Never compiled or run in this
// environment.

import Foundation

/// Creates a new folder under one of `FileAccessPolicy.allowedRoots()`
/// (Desktop, Documents, Downloads, or the dedicated Jarvis folder) — the
/// write counterpart to ListDirectoryTool's read-only listing. Refuses to
/// create anything outside that same allowlist, and refuses to overwrite
/// an existing file/folder silently (fails instead).
enum CreateFolderTool {
    static func make() -> AgentTool {
        AgentTool(name: "create_folder") { input in
            guard let path = input["path"]?.value as? String, !path.trimmingCharacters(in: .whitespaces).isEmpty else {
                return ToolResultPayload(success: false, data: nil, error: "path is required")
            }

            guard let resolved = FileAccessPolicy.resolve(path) else {
                return ToolResultPayload(success: false, data: nil, error: "Path is outside the allowed folders")
            }

            let fileManager = FileManager.default
            if fileManager.fileExists(atPath: resolved.path) {
                return ToolResultPayload(success: false, data: nil, error: "Something already exists at that path: \(path)")
            }

            do {
                try fileManager.createDirectory(at: resolved, withIntermediateDirectories: true)
                let result: [String: AnyCodable] = ["path": AnyCodable(path)]
                return ToolResultPayload(success: true, data: AnyCodable(result), error: nil)
            } catch {
                return ToolResultPayload(success: false, data: nil, error: "Failed to create folder: \(error.localizedDescription)")
            }
        }
    }
}
